source/material_model/simple_nonlinear.cc

```diff
@@ -0,0 +1,371 @@
+          double thermal_expansivity = 0.0;
+          for (unsigned int c=0; c < volume_fractions.size(); ++c)
+            thermal_expansivity += volume_fractions[c] * thermal_expansivities[c];
+
+          // Specific heat at the given positions.
+          double specific_heat = 0.0;
+          for (unsigned int c=0; c < volume_fractions.size(); ++c)
+            specific_heat += volume_fractions[c] * heat_capacity[c];
+
+          // Thermal conductivity at the given positions.
+          double thermal_conductivities = 0.0;
+          for (unsigned int c=0; c < volume_fractions.size(); ++c)
+            thermal_conductivities += volume_fractions[c] * thermal_diffusivity[c] * heat_capacity[c] * densities[c];
+
+          // calculate effective viscosity
+          if (in.strain_rate.size())
+            {
+              // This function calculates viscosities assuming that all the compositional fields
+              // experience the same strain rate (isostrain). Since there is only one process in
+              // this material model (a general powerlaw) we do not need to worry about how to
+              // distribute the strain-rate and stress over the processes.
+              std::vector<double> composition_viscosities(volume_fractions.size());
+              std::vector<SymmetricTensor<2,dim> > composition_viscosities_derivatives(volume_fractions.size());
+
+              for (unsigned int c=0; c < volume_fractions.size(); ++c)
+                {
+                  // If strain rate is zero (like during the first time step) set it to some very small number
+                  // to prevent a division-by-zero, and a floating point exception.
+                  // Otherwise, calculate the square-root of the norm of the second invariant of the deviatoric-
+                  // strain rate (often simplified as epsilondot_ii)
+                  const double edot_ii_strict = std::sqrt(0.5*deviator(in.strain_rate[i])*deviator(in.strain_rate[i]));
+                  const double edot_ii = 2.0 * std::max(edot_ii_strict, min_strain_rate[c] * min_strain_rate[c]);
+
+                  const double stress_exponent_inv = 1/stress_exponent[c];
+                  composition_viscosities[c] = std::max(std::min(std::pow(prefactor[c],-stress_exponent_inv) * std::pow(edot_ii,stress_exponent_inv-1), max_viscosity[c]), min_viscosity[c]);
+                  Assert(dealii::numbers::is_finite(composition_viscosities[c]),ExcMessage ("Error: Viscosity is not finite."));
+                  if (derivatives != NULL)
+                    {
+                      if (edot_ii_strict > min_strain_rate[c] * min_strain_rate[c] && composition_viscosities[c] < max_viscosity[c] && composition_viscosities[c] > min_viscosity[c])
+                        {
+                          //strictly speaking the derivative is this: 0.5 * ((1/stress_exponent)-1) * std::pow(2,2) * out.viscosities[i] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i])
```

Comment:
What do you mean by this comment? Is this a different way to write the line below? Or do you mean the code is wrong? Please clarify that in the comment or remove it. Or is it a question for us?